pkg/apis/clusterregistry/v1alpha1/types.go

```diff
@@ -30,6 +30,62 @@ type Cluster struct {
 	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
 	// +optional
 	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// State is thestate of the cluster. This is not a specification, and is not
+	// meant to be used by actively-reconciling controllers; it is also not
+	// a status, as it contains fields that do not necessarily describe the
+	// status of the cluster, and is not necessarily updated by an active
+	// controller.
+	State ClusterState
+}
+
+// ClusterState contains the state of a cluster.
+type ClusterState struct {
+	KubeAPIServer KubeAPIServer `json:"kube_api_server,omitempty" protobuf:"bytes,1,opt,name=kube_api_server"`
+	AuthInfo      AuthInfo      `json:"auth_info,omitempty" protobuf:"bytes,2,opt,name=auth_info"`
+	CloudProvider CloudProvider `json:"cloud_provider,omitempty" protobuf:"bytes,3,opt,name=cloud_provider"`
+}
+
+type URL string
+
+// KubeAPIServer represents one and only one Kubernetes API server.
+type KubeAPIServer struct {
+	// Servers specifies the addresses of the Kubernetes(s) API server’s network
+	// identity or identities. They can be any valid HTTP URL including the
+	// IP:Port combination or the host name.
+	Servers []URL `json:"servers,omitempty" protobuf:"bytes,1,opt,name=servers"`
+
+	// CABundle is the certificate authority information.
+	CABundle []byte `json:"cabundle,omitempty" protobuf:"bytes,2,opt,name=cabundle"`
+}
+
+// AuthInfo holds public information that describes how a client can get
+// credentials to access the cluster. For example, OAuth2 client registration
+// endpoints and supported flows, Kerberos servers locations, etc.
+//
+// It should not hold any private or sensitive information.
+type AuthInfo struct {
+
+	// AuthProviders is a list of configurations for auth providers.
+	AuthProviders []AuthProviderConfig `json:"auth_providers,omitempty" protobuf:"bytes,1,opt,name=auth_providers"`
+}
+
+// AuthProviderConfig contains the information necessary for a client to
+// authenticate with a Kubernetes API server. It is modeled after
+// k8s.io/client-go/tools/clientcmd/api/v1.AuthProviderConfig.
+type AuthProviderConfig struct {
+	// Name is the name of this configuration.
+	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
+
+	// Config is a map of values that contains the information necessary for a
+	// client to authenticate to a Kubernetes API server.
+	Config map[string]string `json:"config,omitempty" protobuf:"bytes,2,opt,name=config"`
```

Comment:
If this contains secret information, it will mean that `Cluster` is an escalating resource.  I don't think we should couple to k8s secrets, but we can probably factor differently and enable a view of `Cluster` that doesn't give you access to secret values in the Config.

Reply:
I think the intention is that the secret values not be stored in Config at all. The cluster registry is not meant to hold secret values, but it is meant to hold information (e.g., webhook URI) that lets a user authenticate with the cluster in another manner.